lua/plenary/async_lib2/control.lua

```diff
@@ -0,0 +1,209 @@
+local a = require('plenary.async_lib2.async')
+local Deque = require('plenary.async_lib2.structs').Deque
+
+local M = {}
+
+local Condvar = {}
+Condvar.__index = Condvar
+
+---@class Condvar
+---@return Condvar
+function Condvar.new()
+  return setmetatable({handles = {}}, Condvar)
+end
+
+---`blocks` the thread until a notification is received
+Condvar.wait = a.wrap(function(self, callback)
+  -- not calling the callback will block the coroutine
+  table.insert(self.handles, callback)
+end, 2)
+
+---notify everyone that is waiting on this Condvar
+function Condvar:notify_all()
+  if #self.handles == 0 then return end
```

Comment:
I think this check doesn't do anything -- it will short-circuit just the same in the ipairs part of the loop.